File: app/src/main/java/com/yeonsik/fitnessapp/exercise/ExercisePerformanceKey.java
package com.yeonsik.fitnessapp.exercise;

import java.util.Objects;

/**
 * Comparable performance bucket. PR/e1RM/weight statistics must stay inside this exact key.
 */
public final class ExercisePerformanceKey {
    public final String familyId;
    public final String canonicalVariantKey;
    public final LoadState loadState;

    private ExercisePerformanceKey(String familyId, String canonicalVariantKey, LoadState loadState) {
        this.familyId = familyId;
        this.canonicalVariantKey = canonicalVariantKey;
        this.loadState = loadState;
    }

    public static ExercisePerformanceKey of(ExerciseFamilyIdentity identity, LoadState loadState) {
        if (identity == null || !identity.hasVariantIdentity() || loadState == null) {
            return null;
        }
        return new ExercisePerformanceKey(identity.familyId, identity.canonicalVariantKey, loadState);
    }

    public boolean isComparableTo(ExercisePerformanceKey other) {
        return other != null
                && Objects.equals(familyId, other.familyId)
                && Objects.equals(canonicalVariantKey, other.canonicalVariantKey)
                && loadState == other.loadState;
    }

    public String stableValue() {
        return familyId + "|" + canonicalVariantKey + "|" + loadState.id();
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        if (!(other instanceof ExercisePerformanceKey)) return false;
        ExercisePerformanceKey that = (ExercisePerformanceKey) other;
        return Objects.equals(familyId, that.familyId)
                && Objects.equals(canonicalVariantKey, that.canonicalVariantKey)
                && loadState == that.loadState;
    }

    @Override
    public int hashCode() {
        return Objects.hash(familyId, canonicalVariantKey, loadState);
    }

    @Override
    public String toString() {
        return stableValue();
    }
}

File: shared/src/commonMain/kotlin/com/yeonsik/fitness/shared/feature/exercise/model/ExercisePerformanceKey.kt
package com.yeonsik.fitness.shared.feature.exercise.model

import kotlin.jvm.JvmField
import kotlin.jvm.JvmStatic

/** Comparable performance bucket. PR/e1RM/weight statistics stay inside this exact key. */
class ExercisePerformanceKey private constructor(
    @JvmField val familyId: String?,
    @JvmField val canonicalVariantKey: String?,
    @JvmField val loadState: LoadState
) {
    fun isComparableTo(other: ExercisePerformanceKey?): Boolean {
        return other != null
            && familyId == other.familyId
            && canonicalVariantKey == other.canonicalVariantKey
            && loadState == other.loadState
    }

    fun stableValue(): String = "$familyId|$canonicalVariantKey|${loadState.id()}"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ExercisePerformanceKey) return false
        return familyId == other.familyId
            && canonicalVariantKey == other.canonicalVariantKey
            && loadState == other.loadState
    }

    override fun hashCode(): Int {
        var result = 1
        result = 31 * result + (familyId?.hashCode() ?: 0)
        result = 31 * result + (canonicalVariantKey?.hashCode() ?: 0)
        result = 31 * result + loadState.hashCode()
        return result
    }

    override fun toString(): String = stableValue()

    companion object {
        @JvmStatic
        fun of(
            identity: ExerciseFamilyIdentity?,
            loadState: LoadState?
        ): ExercisePerformanceKey? {
            if (identity == null || !identity.hasVariantIdentity() || loadState == null) {
                return null
            }
            return ExercisePerformanceKey(
                identity.familyId,
                identity.canonicalVariantKey,
                loadState
            )
        }
    }
}
